### PR TITLE
Refactor file reading

### PR DESCRIPTION
**Pull request by agateau-gg**
## Description

This PR paves the way to supporting server-side config and reading from various sources. It does so by refactoring the File class and introducing an abstract base class for scannable content, called Scannable.

Scannable has the following API:

- `url` property: Act as a unique identifier for the Scannable. May use custom protocols if required
- `path` property: "path" part of the API, returned as a `pathlib.Path`. Useful for local files
- `is_longer_than(self, size: int) -> bool`: returns true if the length of the *decoded* content is greater than `size`. Used by SecretScanner to decide if afile should be scanned or not
- `content` property: Returns the *decoded* content

File now inherits from Scannable and implements its API.

This PR also introduces StringScannable, which inherits from Scannable too. It is used for scannable content already loaded in memory. Individual patches of a commit are now represented using StringScannable.

## Performance

Interestingly, the performance bench results are very different between my machine and the CI.

CI results:

| command                                       | repository |         prod |      current |   delta |
| --------------------------------------------- | ---------- | -----------: | -----------: | ------: |
| iac scan --exit-zero .                        | Pillow     |  1.58s ±0.10 |  1.54s ±0.04 | -0.04 ≈ |
| iac scan --exit-zero .                        | compose    |  1.45s ±0.02 |  1.42s ±0.13 | -0.03 ≈ |
| iac scan --exit-zero .                        | leaky-repo |  0.97s ±0.04 |  1.01s ±0.07 | +0.04 ≈ |
| iac scan --exit-zero .                        | sqlite     |  0.73s ±0.01 |  0.74s ±0.01 | +0.01 ≈ |
| iac scan --exit-zero .                        | vue        |  1.10s ±0.01 |  1.08s ±0.03 | -0.02 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | Pillow     |  1.90s ±0.11 |  1.94s ±0.03 | +0.04 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | compose    |  1.91s ±0.03 |  1.97s ±0.02 | +0.06 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | leaky-repo |  1.77s ±0.06 |  1.84s ±0.04 | +0.07 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | sqlite     |  0.90s ±0.01 |  0.91s ±0.00 | +0.01 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | vue        |  0.79s ±0.00 |  0.82s ±0.03 | +0.02 ≈ |
| secret scan --exit-zero path -ry .            | Pillow     | 11.70s ±0.27 | 13.85s ±0.22 | +2.14 ▲ |
| secret scan --exit-zero path -ry .            | compose    |  3.50s ±0.01 |  3.52s ±0.17 | +0.01 ≈ |
| secret scan --exit-zero path -ry .            | leaky-repo |  2.29s ±1.02 |  1.94s ±0.14 | -0.35 ≈ |
| secret scan --exit-zero path -ry .            | sqlite     | 26.08s ±0.23 | 27.66s ±0.30 | +1.58 ▲ |
| secret scan --exit-zero path -ry .            | vue        |  5.77s ±0.07 |  5.84s ±0.04 | +0.07 ≈ |

My machine:

| command                                       | repository |         prod |     current |    delta |
| --------------------------------------------- | ---------- | -----------: | ----------: | -------: |
| iac scan --exit-zero .                        | Pillow     |        4.30s |       4.05s |  -0.26 ≈ |
| iac scan --exit-zero .                        | compose    |        2.28s |       2.34s |  +0.06 ≈ |
| iac scan --exit-zero .                        | leaky-repo |        1.74s |       1.76s |  +0.02 ≈ |
| iac scan --exit-zero .                        | sqlite     |        1.28s |       1.45s |  +0.17 ≈ |
| iac scan --exit-zero .                        | vue        |        3.37s |       2.19s |  -1.18 ▼ |
| secret scan --exit-zero commit-range HEAD~9.. | Pillow     |        4.53s |       5.05s |  +0.52 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | compose    |        5.71s |       5.36s |  -0.35 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | leaky-repo |        3.60s |       3.67s |  +0.07 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | sqlite     |        2.60s |       2.38s |  -0.22 ≈ |
| secret scan --exit-zero commit-range HEAD~9.. | vue        |        2.13s |       1.95s |  -0.18 ≈ |
| secret scan --exit-zero path -ry .            | Pillow     | 13.12s ±1.12 | 9.92s ±0.88 |  -3.20 ▼ |
| secret scan --exit-zero path -ry .            | compose    |        5.25s |       4.06s |  -1.19 ▼ |
| secret scan --exit-zero path -ry .            | leaky-repo |        3.92s |       4.10s |  +0.18 ≈ |
| secret scan --exit-zero path -ry .            | sqlite     |       30.08s |      15.57s | -14.52 ▼ |
| secret scan --exit-zero path -ry .            | vue        |        6.84s |       5.02s |  -1.82 ▼ |

I haven't found a rational explanation for this difference. I am going to dig a bit more.

### CODE REVIEW

File: ggshield/scan/scanner.py [hunk 1]
```diff
@@ -245,17 +246,25 @@ def _start_scans(
...
-                chunk.append(file)
+        chunk: List[Scannable] = []
+        for scannable in scannables:
+            if scannable.is_longer_than(DOCUMENT_SIZE_THRESHOLD_BYTES):
+                click.echo(
+                    f"ignoring file over {DOCUMENT_SIZE_THRESHOLD_BYTES // 1024 // 1024} MB:"
+                    f" {scannable.path}",
+                    err=True,
+                )
+                continue
```
Comment (pierrelalanne) [inline ggshield/scan/scanner.py:264]: This scannable won't be part of the skipped_chunk then ?

File: tests/unit/core/test_file_utils.py [hunk 2]
```diff
@@ -34,38 +32,11 @@ def test_generate_files_from_paths(
...
-    assert file.document == expected_content
+    assert file.content == expected_content
 
     assert len(files) == 1
 
 
-@pytest.mark.parametrize(
-    ["filename", "creator"],
-    [
-        ("empty_file", lambda x: x.write_text("")),
```
Comment (pierrelalanne) [inline tests/unit/core/test_file_utils.py:45]: We won't skip empty files anymore, is this expected ?

File: ggshield/scan/scanner.py (hunk 1 above)
Comment (pierrelalanne) [reply ggshield/scan/scanner.py:264]: Edit : I think this gets fixed in another commit.

**Review by pierrelalanne — APPROVED** [review]
No real remarks. The review was tough 😮

File: ggshield/scan/scanner.py (hunk 1 above)
Comment (agateau-gg) [reply ggshield/scan/scanner.py:264]: Yes, sorry for that. The history in this PR could have used more cleanups.

File: tests/unit/core/test_file_utils.py (hunk 2 above)
Comment (agateau-gg) [reply tests/unit/core/test_file_utils.py:45]: We still skip them, but later. The code for this is [here](https://github.com/GitGuardian/ggshield/pull/503/files#diff-5ae06e2f9def4f580d16c631b45c2b6e1fbc9713af8b611d7abd7d67a9ea01d9R266): we only add the Scannable to the chunk if its content is not empty.

**Pull request merged**